akka-stream-tests/src/test/scala/akka/stream/remote/StreamRefsSpec.scala

```diff
@@ -0,0 +1,207 @@
+/**
+ * Copyright (C) 2014-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.remote
+
+import java.nio.file.Paths
+
+import akka.NotUsed
+import akka.actor.{ Actor, ActorRef, ActorSystem, ExtendedActorSystem, Props }
+import akka.serialization._
+import akka.stream.{ ActorMaterializer, remote }
+import akka.stream.remote.scaladsl.{ SinkRef, SourceRef }
+import akka.stream.scaladsl.{ FileIO, Sink, Source }
+import akka.testkit.{ AkkaSpec, ImplicitSender, SocketUtil, TestKit, TestProbe }
+import akka.util.ByteString
+import com.typesafe.config._
+
+import scala.concurrent.duration._
+import scala.concurrent.{ Await, Future }
+
+object StreamRefsSpec {
+
+  object DatasourceActor {
+    def props(probe: ActorRef): Props =
+      Props(new DatasourceActor(probe))
+        .withDispatcher("akka.test.stream-dispatcher")
+  }
+
+  class DatasourceActor(probe: ActorRef) extends Actor {
+    implicit val mat = ActorMaterializer()
+
+    def receive = {
+      //      case "send" ⇒
+      //        /*
+      //         * Here we're able to send a source to a remote recipient
+      //         *
+      //         * For them it's a Source; for us it is a Sink we run data "into"
+      //         */
+      //        val source: Source[String, NotUsed] = Source.single("huge-file-")
+      //        val ref: SourceRef[String] = source.runWith(SourceRef())
+      //
+      //        sender() ! SourceMsg(ref)
+      //
+      //      case "send-bulk" ⇒
+      //        /*
+      //         * Here we're able to send a source to a remote recipient
+      //         * The source is a "bulk transfer one, in which we're ready to send a lot of data"
+      //         *
+      //         * For them it's a Source; for us it is a Sink we run data "into"
+      //         */
+      //        val source: Source[ByteString, NotUsed] = Source.single(ByteString("huge-file-"))
+      //        val ref: SourceRef[ByteString] = source.runWith(SourceRef.bulkTransfer())
+      //        sender() ! BulkSourceMsg(ref)
+
+      case "receive" ⇒
+        /*
+         * We write out code, knowing that the other side will stream the data into it.
+         *
+         * For them it's a Sink; for us it's a Source.
+         */
+        val sink: Future[SinkRef[String]] =
+          SinkRef.source[String]
+            .to(Sink.actorRef(probe, "<COMPLETE>"))
+            .run()
+
+        // FIXME we want to avoid forcing people to do the Future here
+        sender() ! Await.result(sink, 10.seconds)
```

Comment:
This future is a big pain for users, so will have to do something about it I think

Reply:
pipe pattern is normally used for sending an actor message from a future.

Reply:
Yeah, though I really want to get rid of the Future entirely :)

Reply:
Not sure if that's really worth the effort. The difference between `sender() ! sink` and `sink pipeTo sender()` isn't too big...

Reply:
After working with this some more I think it is a pain still -- here it does not look as painful because I don't do the "right thing"™ which user code would be doing -- wrap it in a user message. It gets rather verbose then, since one has to `map(s => MyMsg(id, s)) pipeTo sender()` so the code is not that much of a problem by itself, however I suddenly need to `import dispatcher` AND know about `pipeTo` AND `import` it, also in Java it's not as nice to use...

I have not implemented the way to avoid the Future yet but still think it'd be worth it hmm